contrib/pkg/awsprivatelink/awsprivatelink.go

```diff
@@ -37,3 +66,34 @@ func setLogLevel(cmd *cobra.Command, args []string) {
 		log.SetLevel(log.InfoLevel)
 	}
 }
+
+// Get controller-runtime dynamic client
+func getDynamicClient() client.Client {
+	if err := configv1.Install(scheme.Scheme); err != nil {
+		log.WithError(err).Fatal("Failed to add Openshift configv1 types to the default scheme")
+	}
+
+	dynClient, err := hiveutils.GetClient()
```

Comment:
I feel like this dynamic client reads from the cache but not the API server directly. This leads to a small issue with etcd's optimistic locking. 

For example when I call `hiveutil awsprivatelink endpointvpc add ...` immediately after `hiveutil awsprivatelink enable` (say, in an automated test case), there's a high chance that I got something like:

`time="2023-07-06T15:32:48Z" level=fatal msg="Failed to update HiveConfig/hive" error="Operation cannot be fulfilled on hiveconfigs.hive.openshift.io \"hive\": the object has been modified; please apply your changes to the latest version and try again"`

Currently, after calling `hiveutil awsprivatelink enable` I would call wait.Poll (or something similar) on `hiveutil awsprivatelink endpointvpc add ...` which is idempotent. But is there a better solution ?

Reply:
You're saying you run those two commands serially/synchronously in separate processes and still see this problem? That's... not okay. There should be no way two separate processes are sharing a cache; and the only way I can think of that they're getting confused like this is if they're being assigned to different etcd replicas. IMHO that's an etcd bug.

But that doesn't help you, does it? :innocent: 

In controllers we tend to just ignore 409s like this, as they'll resolve automatically in a subsequent iteration. But in an executable context... ugh. Does it help if you, like, `sleep 5` between those two invocations?

It would be interesting to know how the `oc` command avoids this, as surely it must, or people would be screaming from the rooftops.

Reply:
Dug into controller-runtime's client.New() a bit. Looks like the client actually reads and writes directly from the API server by making requests with a low-level REST client. The only thing that's being cached (as a map) is the resourceMeta (of each GVK requested once by the client), which includes a REST client, a REST mapping etc. but no actual object data.

Reply:
Yep, I ran these two commands serially like so
`cmd := exec.Command("./hiveutil", "awsprivatelink", "enable", ...)`
`out, err = cmd.CombinedOutput()`
...
`cmd = exec.Command("./hiveutil", "awsprivatelink", "endpointvpc", "add", ...)`
`out, err = cmd.CombinedOutput()`
and you are right, sleep for a few seconds help.

Reply:
Yeah this could be caused by etcd not being consistent at that moment. Not entirely sure about this but I think etcd is not strongly consistent for all nodes, as a write operation is considered successful if `quorum` number of nodes acknowledge receipt of the new data.